creator/utils.py

```diff
@@ -0,0 +1,81 @@
+import time
+import logging
+import django_rq
+from rq.job import NoSuchJobError, JobStatus, Job
+from rq.command import send_kill_horse_command
+from rq.worker import Worker, WorkerStatus
+
+logger = logging.getLogger(__name__)
+WAIT_FOR_FAIL_TIME = 0.25
+
+
+def stop_job(job_id, queue=None, delete=True):
+    """
+    Stop a running rq job and optionally delete the job from the rq queue
+    and registries.
+
+    NOTE: This method may not be needed after this
+    https://github.com/rq/rq/issues/1371 is implemented
+
+    For jobs that have failed, do nothing
+    For jobs that are queued or deferred cancel or delete them (see rq.job.Job)
+    For jobs that have started, kill the worker horse executing the job
+
+    If delete is True, remove the job from the redis queue, all rq registries,
+    and from redis
+
+    If delete is False just remove the job from the redis queue
+
+    :param job_id: The redis job id
+    :type job_id: str
+    :param queue: The queue the job was submitted to
+    :type queue: rq.Queue
+    :param delete: whether to delete or cancel the job
+    :type delete: bool
+    """
+    logger.info(f"Preparing to stop redis job {job_id}")
+
+    # Find rq job
+    if not queue:
+        queue = django_rq.get_queue()
+
+    redis_job = None
+    try:
+        redis_job = Job.fetch(job_id, connection=queue.connection)
+    except NoSuchJobError:
+        return
+
+    # Skip jobs that have failed or are finished
+    if (
+        redis_job.get_status() in {JobStatus.FAILED, JobStatus.FINISHED}
+    ):
+        logger.info(f"No redis job: {job_id} found, aborting")
+        return
+
+    # Kill running jobs
+    if redis_job.get_status() == JobStatus.STARTED:
+        for worker in Worker.all(queue=queue):
+            if (worker.state == WorkerStatus.BUSY) and (
+                worker.get_current_job().id == job_id
+            ):
+                logger.info(f"Killing job {job_id}")
+                send_kill_horse_command(queue.connection, worker.name)
+
+    # Wait until job has been moved to failed job registry to delete or
+    # cancel job
+    time.sleep(WAIT_FOR_FAIL_TIME)
```

Comment:
What's the impact of never calling `delete()` or `cancel()` on the job? It seems likely this could happen.

Reply:
If you don't call delete the killed jobs will pile up in the failed job registry. I guess it depends on if you care about that and if you want to differentiate between jobs that failed due to kill horse or something else.  

For cancel, I just realized I need to change the logic bc it removes jobs from the queue that haven't started. So I don't care about waiting for the failed job reg to update